impl/src/metadata.rs

```diff
@@ -194,12 +194,20 @@ fn cargo_generate_lockfile(
 }
 
 pub fn fetch_crate_checksum(index_url: &str, name: &str, version: &str) -> Result<String> {
-  let index = crates_index::BareIndex::from_url(index_url)?;
-
-  let crate_index = index
-    .open_or_clone()?
-    .crate_(name)
-    .ok_or(anyhow!("Failed to find crate '{}' in index", name))?;
+  let crate_index = match url::Url::parse(index_url) {
```

Comment:
This could be regular conditional if you don't actually need the contents of the result.

```let index_path_is_url = url::Url::parse(index_url).is_ok()```

My rule is to use the absolute least powerful expression type that works for the job.

Reply:
I feel like this is actually a valid use for `match` How else can I conditionally set `crate_index` so I don't have to duplicate the logic at the end of this function? Or is that the wrong problem to solve here?

(I'm not deliberately trying to use `match` everywhere, but it does seem to allow me to express more of what I want. I know your gripe with it and am happy to employ alternatives, just need to learn what those are 😅)

Reply:
```rust
let index_path_is_url = url::Url::parse(index_url).is_ok()
let crate_index_path = if index_path_is_url {
  crates_index::Index::new(index_url)
    .crate_(name)
    .ok_or(anyhow!("Failed to find crate '{}' in index", name))?
} else {
  crates_index::BareIndex::from_url(index_url)?
    .open_or_clone()?
    .crate_(name)
    .ok_or(anyhow!("Failed to find crate '{}' in index", name))?
};
```

I don't really feel strongly about this, but it's equivalent.

Reply:
Oh wow, I didn't know you can do that with an `if`. So in every case would you prefer that style over using `match`?

Reply:
I prefer `if` when it suffices. If the match is "doing work", meaning it's serving both as control flow and as destructuring (e.g. Ok(value_i_need)), then I prefer the match. I also prefer the match for enums, or when there's more than two values, unless an early return or series of early returns can be used. Then I prefer the early returns.

Reply:
Updated